source/docs/software/commandbased/commands-v3/structuring-v3-project.rst

```diff
@@ -0,0 +1,465 @@
+     .withPriority(Priorities.EMERGENCY)
+     .named("Emergency Stop");
+
+## Telemetry
+
+Put telemetry in ``periodic()`` methods:
+
+.. code-block:: java
+
+   @Override
+   public void periodic() {
+     SmartDashboard.putNumber("Drivetrain/Left Distance", getLeftDistance());
+     SmartDashboard.putNumber("Drivetrain/Right Distance", getRightDistance());
+     SmartDashboard.putData("Drivetrain/Odometry", odometry);
+
+     // Use mechanism name as prefix for organization
+   }
+
+## Testing
+
+Create test commands for individual mechanisms:
+
+.. code-block:: java
+
+   public class TestCommands {
+
+     public static Command testDrivetrain(Drivetrain drivetrain) {
+       return drivetrain.run(coroutine -> {
+         System.out.println("Testing drivetrain...");
+         drivetrain.tank(0.3, 0.3);
+         coroutine.wait(Seconds.of(1.0));
+         drivetrain.stop();
+         System.out.println("Drivetrain test complete");
+       }).named("Test Drivetrain");
+     }
+
+     public static Command testArm(Arm arm) {
+       return arm.run(coroutine -> {
+         System.out.println("Testing arm...");
+         coroutine.await(arm.moveTo(Constants.ARM_MID));
+         coroutine.wait(Seconds.of(0.5));
+         coroutine.await(arm.moveTo(Constants.ARM_LOW));
+         System.out.println("Arm test complete");
+       }).named("Test Arm");
+     }
+   }
+
+Bind test commands to buttons:
+
+.. code-block:: java
+
+   // In RobotContainer.configureBindings()
+   driver.povUp().onTrue(TestCommands.testDrivetrain(drivetrain));
+   driver.povDown().onTrue(TestCommands.testArm(arm));
+
+## Common Mistakes to Avoid
+
+1. **Scheduling commands inside mechanisms**: Return commands from methods, don't schedule them
+
+.. code-block:: java
+
+   // ❌ BAD: Scheduling inside mechanism
+   public void grab() {
+     run(coro -> { /* ... */ }).schedule();
+   }
+
+   // ✅ GOOD: Return a command
+   public Command grab() {
+     return run(coro -> { /* ... */ }).named("Grab");
+   }
+
+2. **Forgetting to call yield()**: Always yield inside loops
+
+.. code-block:: java
+
+   // ❌ BAD: No yield
+   while (condition) {
+     doWork();
+   }
+
+   // ✅ GOOD: Yields every cycle
+   while (condition) {
+     doWork();
+     coroutine.yield();
+   }
+
+3. **Not naming commands**: All commands need names
+
+.. code-block:: java
+
+   // ❌ BAD: Missing name
+   Command cmd = mechanism.run(coro -> { /* ... */ });
+
+   // ✅ GOOD: Has name
+   Command cmd = mechanism.run(coro -> { /* ... */ }).named("Action");
+
+4. **Exposing hardware objects**: Keep hardware private
+
+.. code-block:: java
+
+   // ❌ BAD: Public motor
+   public TalonFX motor = new TalonFX(1);
+
+   // ✅ GOOD: Private hardware, public methods
+   private final TalonFX motor = new TalonFX(1);
+
+   public void setSpeed(double speed) {
+     motor.set(speed);
+   }
+
+5. **Complex logic in RobotContainer**: Move complex commands to separate files
```

Comment:
`RobotContainer` isn't recommended in either v2 or v3 any more. This section could also use examples

Reply:
I know this has been discussed but I don't think we have ever finalized this recommendation.  We should discuss this as a team.  We still include RobotContainer in all of our templates many example and mention it in many places so if we want to commit to this I just want to make sure everyone is on board and that we update everything else to have a consistent message to teams cause this is a large deviation from current practice.